File: 2021/08/29/longest_palindrome.py
"""
https://leetcode.com/problems/longest-palindrome/
Author: Andrew Jarombek
Date: 8/29/2021
"""


def longest_palindrome(s: str) -> int:
    char_dict = {}

    for c in s:
        char_dict[c] = char_dict[c] + 1 if c in char_dict else 1

    items = list(char_dict.values())
    items.sort(reverse=True)

    odd_used = False
    length = 0

    for item in items:
        if item % 2 == 0:
            length += item
        else:
            if not odd_used:
                length += item
                odd_used = True
            else:
                length += item - 1

    return length


if __name__ == '__main__':
    assert longest_palindrome("abccccdd") == 7

File: 2022/07/21/java/LongestPalindrome.java
import java.util.HashMap;
import java.util.Map;

/**
 * https://leetcode.com/problems/longest-palindrome/
 * @author Andrew Jarombek
 * @since 7/21/2022
 */

public class LongestPalindrome {
    public static int longestPalindrome(String s) {
        Map<Character, Integer> map = new HashMap<>();

        for (char c : s.toCharArray()) {
            map.put(c, map.getOrDefault(c, 0) + 1);
        }

        int result = 0;
        boolean odd = false;

        for (int value : map.values()) {
            if (value % 2 != 0)
                odd = true;

            result += value - (value % 2);
        }

        if (odd)
            result++;

        return result;
    }

    public static void main(String[] args) {
        assert longestPalindrome("abccccdd") == 7;
    }
}
